File: examples/physics2/main.lua
-- love2d Physics Phase 2 example
-- Demonstrates: WheelJoint (car), MotorJoint (tracker), joint anchors,
-- reaction force, contact point + approach speed
-- Hard-coded motor offset for stabilization comparison

local world
local ground, rampBody
local chassis, wheelL, wheelR
local wheelJointL, wheelJointR
local trackerAnchor, trackerBody, motorJoint

local GROUND_Y = 450
local CAR_X = 200
local CAR_Y = 380

-- Hard-coded motor target (absolute position)
local TARGET_X = 700
local TARGET_Y = 350

-- Contact flash at exact point with intensity
local flashes = {}

-- Track key state for HUD
local driving = 0

-- Elapsed timer
local elapsed = 0

function love.load()
  love.window.setTitle("Physics Phase 2 — love2d")
  love.window.setMode(800, 600)
  love.graphics.setBackgroundColor(20/255, 20/255, 30/255)
  love.physics.setMeter(30)

  world = love.physics.newWorld(0, 9.81 * 30, true)

  -- Contact callback with point + speed (postSolve is 4th callback)
  world:setCallbacks(nil, nil, nil, function(a, b, contact, ni1, ti1, ni2, ti2)
    local x1, y1 = contact:getPositions()
    if x1 and y1 then
      local speed = math.abs(ni1 or 0)
      if speed > 0.5 then
        table.insert(flashes, { x = x1, y = y1, speed = math.min(speed, 20), timer = 0.3 })
      end
    end
  end)

  -- Ground (flat)
  ground = love.physics.newBody(world, 400, GROUND_Y, "static")
  local groundShape = love.physics.newRectangleShape(800, 20)
  love.physics.newFixture(ground, groundShape, 1)

  -- Ramp
  rampBody = love.physics.newBody(world, 500, GROUND_Y - 30, "static")
  local rampShape = love.physics.newPolygonShape(-80, 30, 80, 30, 80, -10)
  local rampFix = love.physics.newFixture(rampBody, rampShape, 1)
  rampFix:setFriction(0.8)

  -- === Car with WheelJoints ===
  -- Chassis
  chassis = love.physics.newBody(world, CAR_X, CAR_Y, "dynamic")
  local chassisShape = love.physics.newRectangleShape(80, 20)
  local chassisFix = love.physics.newFixture(chassis, chassisShape, 2)
  chassisFix:setFriction(0.3)

  -- Left wheel
  wheelL = love.physics.newBody(world, CAR_X - 30, CAR_Y + 20, "dynamic")
  local wheelShapeL = love.physics.newCircleShape(12)
  local wheelFixL = love.physics.newFixture(wheelL, wheelShapeL, 1)
  wheelFixL:setFriction(1.0)

  -- Right wheel
  wheelR = love.physics.newBody(world, CAR_X + 30, CAR_Y + 20, "dynamic")
  local wheelShapeR = love.physics.newCircleShape(12)
  local wheelFixR = love.physics.newFixture(wheelR, wheelShapeR, 1)
  wheelFixR:setFriction(1.0)

  -- Wheel joints (axis = vertical for suspension)
  wheelJointL = love.physics.newWheelJoint(chassis, wheelL, CAR_X - 30, CAR_Y + 20, 0, 1, false)
  wheelJointL:setSpringFrequency(4.0)
  wheelJointL:setSpringDampingRatio(0.7)
  wheelJointL:setMaxMotorTorque(1000)
  wheelJointL:setMotorSpeed(0)
  wheelJointL:setMotorEnabled(true)

  wheelJointR = love.physics.newWheelJoint(chassis, wheelR, CAR_X + 30, CAR_Y + 20, 0, 1, false)
  wheelJointR:setSpringFrequency(4.0)
  wheelJointR:setSpringDampingRatio(0.7)
  wheelJointR:setMaxMotorTorque(1000)
  wheelJointR:setMotorSpeed(0)
  wheelJointR:setMotorEnabled(true)

  -- === Motor Joint (tracker) ===
  trackerAnchor = love.physics.newBody(world, 600, 200, "static")

  trackerBody = love.physics.newBody(world, 600, 200, "dynamic")
  trackerBody:setGravityScale(0)
  local trackerShape = love.physics.newRectangleShape(40, 40)
  local trackerFix = love.physics.newFixture(trackerBody, trackerShape, 1)
  trackerFix:setFriction(0.3)

  motorJoint = love.physics.newMotorJoint(trackerAnchor, trackerBody, 0.3)
  motorJoint:setMaxForce(500)
  motorJoint:setMaxTorque(200)

  -- Set hard-coded offset immediately
  local ax, ay = trackerAnchor:getPosition()
  motorJoint:setLinearOffset(TARGET_X - ax, TARGET_Y - ay)
end

function love.update(dt)
  elapsed = elapsed + dt

  -- Drive car with arrow keys
  driving = 0
  if love.keyboard.isDown("right") then driving = 1 end
  if love.keyboard.isDown("left") then driving = -1 end
  wheelJointL:setMotorSpeed(driving * 15)
  wheelJointR:setMotorSpeed(driving * 15)

  -- Motor joint offset is fixed — no mouse tracking

  world:update(dt)

  -- Fade flashes
  for i = #flashes, 1, -1 do
    flashes[i].timer = flashes[i].timer - dt
    if flashes[i].timer <= 0 then
      table.remove(flashes, i)
    end
  end
end

function love.draw()
  -- === Ground & Ramp ===
  love.graphics.setColor(80/255, 80/255, 80/255)
  local gx, gy = ground:getPosition()
  love.graphics.rectangle("fill", gx - 400, gy - 10, 800, 20)

  -- Ramp (draw as polygon)
  love.graphics.setColor(100/255, 80/255, 60/255)
  love.graphics.push()
  love.graphics.translate(rampBody:getPosition())
  love.graphics.polygon("fill", -80, 30, 80, 30, 80, -10)
  love.graphics.pop()

  -- === Car ===
  -- Chassis
  love.graphics.setColor(100/255, 150/255, 220/255)
  love.graphics.push()
  love.graphics.translate(chassis:getPosition())
  love.graphics.rotate(chassis:getAngle())
  love.graphics.rectangle("fill", -40, -10, 80, 20)
  love.graphics.pop()

  -- Wheels
  love.graphics.setColor(60/255, 60/255, 60/255)
  local wlx, wly = wheelL:getPosition()
  love.graphics.circle("fill", wlx, wly, 12)
  local wrx, wry = wheelR:getPosition()
  love.graphics.circle("fill", wrx, wry, 12)

  -- Wheel spokes (show rotation)
  love.graphics.setColor(200/255, 200/255, 200/255)
  local alL = wheelL:getAngle()
  love.graphics.line(wlx, wly, wlx + math.cos(alL) * 10, wly + math.sin(alL) * 10)
  local alR = wheelR:getAngle()
  love.graphics.line(wrx, wry, wrx + math.cos(alR) * 10, wry + math.sin(alR) * 10)

  -- Joint anchors (small dots)
  love.graphics.setColor(1, 1, 0)
  local a1x, a1y, b1x, b1y = wheelJointL:getAnchors()
  love.graphics.circle("fill", a1x, a1y, 3)
  love.graphics.circle("fill", b1x, b1y, 3)
  local a2x, a2y, b2x, b2y = wheelJointR:getAnchors()
  love.graphics.circle("fill", a2x, a2y, 3)
  love.graphics.circle("fill", b2x, b2y, 3)

  -- Suspension lines (anchor A to anchor B)
  love.graphics.setColor(1, 1, 0, 0.5)
  love.graphics.line(a1x, a1y, b1x, b1y)
  love.graphics.line(a2x, a2y, b2x, b2y)

  -- === Motor Joint Tracker ===
  -- Target crosshair (fixed position)
  love.graphics.setColor(1, 0.4, 0.4, 0.5)
  love.graphics.circle("line", TARGET_X, TARGET_Y, 15)
  love.graphics.line(TARGET_X - 10, TARGET_Y, TARGET_X + 10, TARGET_Y)
  love.graphics.line(TARGET_X, TARGET_Y - 10, TARGET_X, TARGET_Y + 10)

  -- Tracker body
  love.graphics.setColor(220/255, 100/255, 100/255)
  love.graphics.push()
  love.graphics.translate(trackerBody:getPosition())
  love.graphics.rotate(trackerBody:getAngle())
  love.graphics.rectangle("fill", -20, -20, 40, 40)
  love.graphics.pop()

  -- Motor joint anchor
  love.graphics.setColor(1, 0.8, 0)
  local ma1x, ma1y, ma2x, ma2y = motorJoint:getAnchors()
  love.graphics.circle("fill", ma1x, ma1y, 4)
  love.graphics.circle("fill", ma2x, ma2y, 4)
  love.graphics.line(ma1x, ma1y, ma2x, ma2y)

  -- === Contact flashes ===
  for _, flash in ipairs(flashes) do
    local alpha = flash.timer / 0.3
    local r = math.min(flash.speed / 5, 4) + 3
    love.graphics.setColor(1, 1, 0.5, alpha)
    love.graphics.circle("fill", flash.x, flash.y, r)
  end

  -- === HUD ===
  love.graphics.setColor(200/255, 200/255, 200/255)
  love.graphics.print("Physics Phase 2 — WheelJoint + MotorJoint", 10, 10)
  love.graphics.print("Arrow keys: drive car   |   R: reset   |   ESC: quit", 10, 30)

  -- Timer
  love.graphics.setColor(1, 1, 0.4)
  love.graphics.print(string.format("Time: %.2fs", elapsed), 10, 50)

  -- Tracker distance to target
  local tbx, tby = trackerBody:getPosition()
  local dist = math.sqrt((tbx - TARGET_X)^2 + (tby - TARGET_Y)^2)
  love.graphics.print(string.format("Tracker dist to target: %.1f px", dist), 10, 70)

  -- Reaction force on left wheel joint (inverseDt = 60 for 1/60s timestep)
  love.graphics.setColor(200/255, 200/255, 200/255)
  local rfx, rfy = wheelJointL:getReactionForce(60)
  love.graphics.print(string.format("Left wheel reaction: (%.1f, %.1f)", rfx, rfy), 10, 90)

  -- Spring info
  love.graphics.print(string.format("Spring: %.1f Hz, damping %.1f",
    wheelJointL:getSpringFrequency(), wheelJointL:getSpringDampingRatio()), 10, 110)

  -- Motor joint offset
  local ox, oy = motorJoint:getLinearOffset()
  love.graphics.print(string.format("Motor offset: (%.0f, %.0f)", ox, oy), 10, 130)

  -- Driving indicator
  local driveText = "IDLE"
  if driving > 0 then driveText = ">>> RIGHT >>>"
  elseif driving < 0 then driveText = "<<< LEFT <<<"
  end
  love.graphics.print("Drive: " .. driveText, 10, 150)

  love.graphics.setColor(1, 1, 1)
  love.graphics.print("FPS: " .. love.timer.getFPS(), 700, 570)
end

function love.keypressed(key)
  if key == "escape" then
    love.event.quit()
  elseif key == "r" then
    elapsed = 0
    chassis:setPosition(CAR_X, CAR_Y)
    chassis:setAngle(0)
    chassis:setLinearVelocity(0, 0)
    chassis:setAngularVelocity(0)
    wheelL:setPosition(CAR_X - 30, CAR_Y + 20)
    wheelL:setAngle(0)
    wheelL:setLinearVelocity(0, 0)
    wheelL:setAngularVelocity(0)
    wheelR:setPosition(CAR_X + 30, CAR_Y + 20)
    wheelR:setAngle(0)
    wheelR:setLinearVelocity(0, 0)
    wheelR:setAngularVelocity(0)
    -- Reset tracker
    trackerBody:setPosition(600, 200)
    trackerBody:setLinearVelocity(0, 0)
    trackerBody:setAngularVelocity(0)
  end
end

File: examples/physics2/main.ts
// jove2d Physics Phase 2 example
// Demonstrates: WheelJoint (car), MotorJoint (tracker), joint anchors,
// reaction force, contact point + approach speed
// Hard-coded motor offset for stabilization comparison

import jove from "../../src/index.ts";

let world: ReturnType<typeof jove.physics.newWorld>;
let ground: ReturnType<typeof jove.physics.newBody>;
let rampBody: ReturnType<typeof jove.physics.newBody>;
let chassis: ReturnType<typeof jove.physics.newBody>;
let wheelL: ReturnType<typeof jove.physics.newBody>;
let wheelR: ReturnType<typeof jove.physics.newBody>;
let wheelJointL: ReturnType<typeof jove.physics.newWheelJoint>;
let wheelJointR: ReturnType<typeof jove.physics.newWheelJoint>;
let trackerAnchor: ReturnType<typeof jove.physics.newBody>;
let trackerBody: ReturnType<typeof jove.physics.newBody>;
let motorJoint: ReturnType<typeof jove.physics.newMotorJoint>;

const GROUND_Y = 450;
const CAR_X = 200;
const CAR_Y = 380;

// Hard-coded motor target offset (relative to anchor at 600,200)
const TARGET_X = 700;
const TARGET_Y = 350;

// Contact flash at exact point with intensity
const flashes: { x: number; y: number; speed: number; timer: number }[] = [];

// Track driving state for HUD
let driving = 0;

// Elapsed timer
let elapsed = 0;

const FIXED_DT = 1 / 60;
let accumulator = 0;

await jove.run({
  load() {
    jove.window.setTitle("Physics Phase 2 — jove2d");
    jove.graphics.setBackgroundColor(20, 20, 30);

    jove.physics.setMeter(30);
    world = jove.physics.newWorld(0, 9.81 * 30);

    // Contact callback with point + speed
    world.setCallbacks({
      postSolve(contact, normalImpulse, _tangentImpulse) {
        const [px, py] = contact.getPositions();
        const speed = contact.getNormalImpulse();
        if (speed > 0.5) {
          flashes.push({ x: px, y: py, speed: Math.min(speed, 20), timer: 0.3 });
        }
      },
    });

    // Ground (flat)
    ground = jove.physics.newBody(world, 400, GROUND_Y, "static");
    const groundShape = jove.physics.newRectangleShape(800, 20);
    jove.physics.newFixture(ground, groundShape, 1);

    // Ramp
    rampBody = jove.physics.newBody(world, 500, GROUND_Y - 30, "static");
    const rampShape = jove.physics.newPolygonShape(-80, 30, 80, 30, 80, -10);
    const rampFix = jove.physics.newFixture(rampBody, rampShape, 1);
    rampFix.setFriction(0.8);

    // === Car with WheelJoints ===
    // Chassis
    chassis = jove.physics.newBody(world, CAR_X, CAR_Y, "dynamic");
    const chassisShape = jove.physics.newRectangleShape(80, 20);
    const chassisFix = jove.physics.newFixture(chassis, chassisShape, 2);
    chassisFix.setFriction(0.3);

    // Left wheel
    wheelL = jove.physics.newBody(world, CAR_X - 30, CAR_Y + 20, "dynamic");
    const wheelShapeL = jove.physics.newCircleShape(12);
    const wheelFixL = jove.physics.newFixture(wheelL, wheelShapeL, 1);
    wheelFixL.setFriction(1.0);

    // Right wheel
    wheelR = jove.physics.newBody(world, CAR_X + 30, CAR_Y + 20, "dynamic");
    const wheelShapeR = jove.physics.newCircleShape(12);
    const wheelFixR = jove.physics.newFixture(wheelR, wheelShapeR, 1);
    wheelFixR.setFriction(1.0);

    // Wheel joints (axis = vertical for suspension)
    wheelJointL = jove.physics.newWheelJoint(chassis, wheelL, CAR_X - 30, CAR_Y + 20, 0, 1);
    wheelJointL.setSpringEnabled(true);
    wheelJointL.setSpringFrequency(4.0);
    wheelJointL.setSpringDampingRatio(0.7);
    wheelJointL.setMaxMotorTorque(1000);
    wheelJointL.setMotorSpeed(0);
    wheelJointL.setMotorEnabled(true);

    wheelJointR = jove.physics.newWheelJoint(chassis, wheelR, CAR_X + 30, CAR_Y + 20, 0, 1);
    wheelJointR.setSpringEnabled(true);
    wheelJointR.setSpringFrequency(4.0);
    wheelJointR.setSpringDampingRatio(0.7);
    wheelJointR.setMaxMotorTorque(1000);
    wheelJointR.setMotorSpeed(0);
    wheelJointR.setMotorEnabled(true);

    // === Motor Joint (tracker) ===
    trackerAnchor = jove.physics.newBody(world, 600, 200, "static");

    trackerBody = jove.physics.newBody(world, 600, 200, "dynamic");
    trackerBody.setGravityScale(0); // no gravity for tracker
    const trackerShape = jove.physics.newRectangleShape(40, 40);
    const trackerFix = jove.physics.newFixture(trackerBody, trackerShape, 1);
    trackerFix.setFriction(0.3);

    motorJoint = jove.physics.newMotorJoint(trackerAnchor, trackerBody, 0.3);
    motorJoint.setMaxForce(500);
    motorJoint.setMaxTorque(200);

    // Set hard-coded offset immediately
    const [ax, ay] = trackerAnchor.getPosition();
    motorJoint.setLinearOffset(TARGET_X - ax, TARGET_Y - ay);
  },

  update(dt) {
    elapsed += dt;

    // Drive car with arrow keys
    driving = 0;
    if (jove.keyboard.isDown("right")) driving = 1;
    if (jove.keyboard.isDown("left")) driving = -1;
    wheelJointL.setMotorSpeed(driving * 15);
    wheelJointR.setMotorSpeed(driving * 15);

    accumulator += dt;
    while (accumulator >= FIXED_DT) {
      world.update(FIXED_DT);
      accumulator -= FIXED_DT;
    }

    // Fade flashes
    for (let i = flashes.length - 1; i >= 0; i--) {
      flashes[i].timer -= dt;
      if (flashes[i].timer <= 0) flashes.splice(i, 1);
    }
  },

  draw() {
    // === Ground & Ramp ===
    jove.graphics.setColor(80, 80, 80);
    const [gx, gy] = ground.getPosition();
    jove.graphics.rectangle("fill", gx - 400, gy - 10, 800, 20);

    // Ramp (draw as polygon)
    jove.graphics.setColor(100, 80, 60);
    jove.graphics.push();
    const [rx, ry] = rampBody.getPosition();
    jove.graphics.translate(rx, ry);
    jove.graphics.polygon("fill", -80, 30, 80, 30, 80, -10);
    jove.graphics.pop();

    // === Car ===
    // Chassis
    jove.graphics.setColor(100, 150, 220);
    jove.graphics.push();
    const [cx, cy] = chassis.getPosition();
    jove.graphics.translate(cx, cy);
    jove.graphics.rotate(chassis.getAngle());
    jove.graphics.rectangle("fill", -40, -10, 80, 20);
    jove.graphics.pop();

    // Wheels
    jove.graphics.setColor(60, 60, 60);
    const [wlx, wly] = wheelL.getPosition();
    jove.graphics.circle("fill", wlx, wly, 12);
    const [wrx, wry] = wheelR.getPosition();
    jove.graphics.circle("fill", wrx, wry, 12);

    // Wheel spokes (show rotation)
    jove.graphics.setColor(200, 200, 200);
    const alL = wheelL.getAngle();
    jove.graphics.line(wlx, wly, wlx + Math.cos(alL) * 10, wly + Math.sin(alL) * 10);
    const alR = wheelR.getAngle();
    jove.graphics.line(wrx, wry, wrx + Math.cos(alR) * 10, wry + Math.sin(alR) * 10);

    // Joint anchors (small dots)
    jove.graphics.setColor(255, 255, 0);
    const [a1x, a1y] = wheelJointL.getAnchorA();
    const [b1x, b1y] = wheelJointL.getAnchorB();
    jove.graphics.circle("fill", a1x, a1y, 3);
    jove.graphics.circle("fill", b1x, b1y, 3);
    const [a2x, a2y] = wheelJointR.getAnchorA();
    const [b2x, b2y] = wheelJointR.getAnchorB();
    jove.graphics.circle("fill", a2x, a2y, 3);
    jove.graphics.circle("fill", b2x, b2y, 3);

    // Suspension lines (anchor A to anchor B)
    jove.graphics.setColor(255, 255, 0, 128);
    jove.graphics.line(a1x, a1y, b1x, b1y);
    jove.graphics.line(a2x, a2y, b2x, b2y);

    // === Motor Joint Tracker ===
    // Target crosshair (fixed position)
    jove.graphics.setColor(255, 100, 100, 128);
    jove.graphics.circle("line", TARGET_X, TARGET_Y, 15);
    jove.graphics.line(TARGET_X - 10, TARGET_Y, TARGET_X + 10, TARGET_Y);
    jove.graphics.line(TARGET_X, TARGET_Y - 10, TARGET_X, TARGET_Y + 10);

    // Tracker body
    jove.graphics.setColor(220, 100, 100);
    jove.graphics.push();
    const [tx, ty] = trackerBody.getPosition();
    jove.graphics.translate(tx, ty);
    jove.graphics.rotate(trackerBody.getAngle());
    jove.graphics.rectangle("fill", -20, -20, 40, 40);
    jove.graphics.pop();

    // Motor joint anchors
    jove.graphics.setColor(255, 200, 0);
    const [ma1x, ma1y] = motorJoint.getAnchorA();
    const [ma2x, ma2y] = motorJoint.getAnchorB();
    jove.graphics.circle("fill", ma1x, ma1y, 4);
    jove.graphics.circle("fill", ma2x, ma2y, 4);
    jove.graphics.line(ma1x, ma1y, ma2x, ma2y);

    // === Contact flashes ===
    for (const flash of flashes) {
      const alpha = Math.floor((flash.timer / 0.3) * 255);
      const r = Math.min(flash.speed / 5, 4) + 3;
      jove.graphics.setColor(255, 255, 128, alpha);
      jove.graphics.circle("fill", flash.x, flash.y, r);
    }

    // === HUD ===
    jove.graphics.setColor(200, 200, 200);
    jove.graphics.print("Physics Phase 2 — WheelJoint + MotorJoint", 10, 10);
    jove.graphics.print("Arrow keys: drive car   |   R: reset   |   ESC: quit", 10, 30);

    // Timer
    jove.graphics.setColor(255, 255, 100);
    jove.graphics.print(`Time: ${elapsed.toFixed(2)}s`, 10, 50);

    // Tracker distance to target
    const [tbx, tby] = trackerBody.getPosition();
    const dist = Math.sqrt((tbx - TARGET_X) ** 2 + (tby - TARGET_Y) ** 2);
    jove.graphics.print(`Tracker dist to target: ${dist.toFixed(1)} px`, 10, 70);

    // Reaction force on left wheel joint
    jove.graphics.setColor(200, 200, 200);
    const [rfx, rfy] = wheelJointL.getReactionForce(1 / 60);
    jove.graphics.print(`Left wheel reaction: (${rfx.toFixed(1)}, ${rfy.toFixed(1)})`, 10, 90);

    // Spring info
    jove.graphics.print(`Spring: ${wheelJointL.getSpringFrequency().toFixed(1)} Hz, damping ${wheelJointL.getSpringDampingRatio().toFixed(1)}`, 10, 110);

    // Motor joint offset
    const [ox, oy] = motorJoint.getLinearOffset();
    jove.graphics.print(`Motor offset: (${ox.toFixed(0)}, ${oy.toFixed(0)})`, 10, 130);

    // Driving indicator
    let driveText = "IDLE";
    if (driving > 0) driveText = ">>> RIGHT >>>";
    else if (driving < 0) driveText = "<<< LEFT <<<";
    jove.graphics.print(`Drive: ${driveText}`, 10, 150);

    jove.graphics.setColor(255, 255, 255);
    jove.graphics.print(`FPS: ${jove.timer.getFPS()}`, 700, 570);
  },

  keypressed(key) {
    if (key === "escape") jove.event.quit();
    if (key === "r") {
      elapsed = 0;
      // Reset car position
      chassis.setPosition(CAR_X, CAR_Y);
      chassis.setAngle(0);
      chassis.setLinearVelocity(0, 0);
      chassis.setAngularVelocity(0);
      wheelL.setPosition(CAR_X - 30, CAR_Y + 20);
      wheelL.setAngle(0);
      wheelL.setLinearVelocity(0, 0);
      wheelL.setAngularVelocity(0);
      wheelR.setPosition(CAR_X + 30, CAR_Y + 20);
      wheelR.setAngle(0);
      wheelR.setLinearVelocity(0, 0);
      wheelR.setAngularVelocity(0);
      // Reset tracker
      trackerBody.setPosition(600, 200);
      trackerBody.setLinearVelocity(0, 0);
      trackerBody.setAngularVelocity(0);
    }
  },
});
